File: Sample Apps/DotNet Custom Attribute/Core/WebApplicationCore8/WebApplicationCore8/Util/MotadataDynamicInstrumentation.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;

namespace WebApplicationCore8.Util
{
    /// <summary>
    /// Utility for adding custom attributes (tags) to the current OpenTelemetry span in .NET.
    /// All attribute keys are automatically prefixed with "apm.".
    ///
    /// Supported types:
    /// bool, double, float, int, long, string
    /// bool[], double[], float[], int[], long[], string[]
    ///
    /// This class is thread-safe and exception-safe.
    /// All methods silently ignore null/invalid inputs and will never throw exceptions.
    /// Every operation is wrapped in try-catch for maximum robustness.
    /// </summary>
    public static class MotadataDynamicInstrumentation
    {
        private const string DefaultPrefix = "apm.";

        private static string PrefixKey(string key)
        {
            try
            {
                if (string.IsNullOrEmpty(key))
                    return null;

                return key.StartsWith(DefaultPrefix, StringComparison.Ordinal)
                    ? key
                    : DefaultPrefix + key;
            }
            catch
            {
                return null;
            }
        }

        private static Activity CurrentActivity
        {
            get
            {
                try
                {
                    return Activity.Current;
                }
                catch
                {
                    return null;
                }
            }
        }

        // ==================== Primitive types (now fully try-catch protected) ====================

        public static void Set(string key, bool value)
        {
            try
            {
                if (string.IsNullOrEmpty(key) || CurrentActivity == null)
                    return;

                string prefixed = PrefixKey(key);
                if (prefixed == null)
                    return;

                CurrentActivity.SetTag(prefixed, value);
            }
            catch
            {
                // Swallow all exceptions
            }
        }

        public static void Set(string key, double value)
        {
            try
            {
                if (string.IsNullOrEmpty(key) || CurrentActivity == null)
                    return;

                string prefixed = PrefixKey(key);
                if (prefixed == null)
                    return;

                CurrentActivity.SetTag(prefixed, value);
            }
            catch
            {
                // Swallow all exceptions
            }
        }

        public static void Set(string key, float value)
        {
            try
            {
                if (string.IsNullOrEmpty(key) || CurrentActivity == null)
                    return;

                string prefixed = PrefixKey(key);
                if (prefixed == null)
                    return;

                CurrentActivity.SetTag(prefixed, value);
            }
            catch
            {
                // Swallow all exceptions
            }
        }

        public static void Set(string key, int value)
        {
            try
            {
                if (string.IsNullOrEmpty(key) || CurrentActivity == null)
                    return;

                string prefixed = PrefixKey(key);
                if (prefixed == null)
                    return;

                CurrentActivity.SetTag(prefixed, value);
            }
            catch
            {
                // Swallow all exceptions
            }
        }

        public static void Set(string key, long value)
        {
            try
            {
                if (string.IsNullOrEmpty(key) || CurrentActivity == null)
                    return;

                string prefixed = PrefixKey(key);
                if (prefixed == null)
                    return;

                CurrentActivity.SetTag(prefixed, value);
            }
            catch
            {
                // Swallow all exceptions
            }
        }

        public static void Set(string key, string value)
        {
            try
            {
                if (string.IsNullOrEmpty(key) || value == null || CurrentActivity == null)
                    return;

                string prefixed = PrefixKey(key);
                if (prefixed == null)
                    return;

                CurrentActivity.SetTag(prefixed, value);
            }
            catch
            {
                // Swallow all exceptions
            }
        }

        // ==================== Array types (already protected, kept consistent) ====================

        public static void SetBoolArray(string key, bool[] values)
        {
            try
            {
                if (string.IsNullOrEmpty(key) || values == null || values.Length == 0 || CurrentActivity == null)
                    return;

                string prefixed = PrefixKey(key);
                if (prefixed == null)
                    return;

                CurrentActivity.SetTag(prefixed, values);
            }
            catch
            {
                // Ignore all exceptions
            }
        }

        public static void SetDoubleArray(string key, double[] values)
        {
            try
            {
                if (string.IsNullOrEmpty(key) || values == null || values.Length == 0 || CurrentActivity == null)
                    return;

                string prefixed = PrefixKey(key);
                if (prefixed == null)
                    return;

                CurrentActivity.SetTag(prefixed, values);
            }
            catch
            {
                // Ignore all exceptions
            }
        }

        public static void SetFloatArray(string key, float[] values)
        {
            try
            {
                if (string.IsNullOrEmpty(key) || values == null || values.Length == 0 || CurrentActivity == null)
                    return;

                string prefixed = PrefixKey(key);
                if (prefixed == null)
                    return;

                CurrentActivity.SetTag(prefixed, values);
            }
            catch
            {
                // Ignore all exceptions
            }
        }

        public static void SetIntArray(string key, int[] values)
        {
            try
            {
                if (string.IsNullOrEmpty(key) || values == null || values.Length == 0 || CurrentActivity == null)
                    return;

                string prefixed = PrefixKey(key);
                if (prefixed == null)
                    return;

                CurrentActivity.SetTag(prefixed, values);
            }
            catch
            {
                // Ignore all exceptions
            }
        }

        public static void SetLongArray(string key, long[] values)
        {
            try
            {
                if (string.IsNullOrEmpty(key) || values == null || values.Length == 0 || CurrentActivity == null)
                    return;

                string prefixed = PrefixKey(key);
                if (prefixed == null)
                    return;

                CurrentActivity.SetTag(prefixed, values);
            }
            catch
            {
                // Ignore all exceptions
            }
        }

        public static void SetStringArray(string key, string[] values)
        {
            try
            {
                if (string.IsNullOrEmpty(key) || values == null || values.Length == 0 || CurrentActivity == null)
                    return;

                string prefixed = PrefixKey(key);
                if (prefixed == null)
                    return;

                // Manual null filtering (no LINQ, maximum compatibility)
                var cleaned = new List<string>(values.Length);
                foreach (var s in values)
                {
                    if (s != null)
                        cleaned.Add(s);
                }

                if (cleaned.Count == 0)
                    return;

                CurrentActivity.SetTag(prefixed, cleaned.ToArray());
            }
            catch
            {
                // Ignore all exceptions
            }
        }
    }
}

File: Sample Apps/Python Custom Attribute/otelapi/util/MotadataDynamicInstrumentation.py
from __future__ import annotations

from typing import Optional, Sequence
from opentelemetry import trace


class MotadataDynamicInstrumentation:
    """
    Simple utility to add custom attributes to the current OpenTelemetry span.
    Keys are automatically prefixed with 'apm.' if not already present.

    Never raises exceptions. Silently ignores everything invalid.
    """

    _PREFIX = "apm."

    def __init__(self) -> None:
        raise RuntimeError("This is a static utility class")

    @staticmethod
    def _add_prefix(key: Optional[str]) -> Optional[str]:
        if not key:
            return None
        try:
            return key if key.startswith(MotadataDynamicInstrumentation._PREFIX) else MotadataDynamicInstrumentation._PREFIX + key
        except Exception:
            return None

    @staticmethod
    def _try_set_attribute(key: str, value: any) -> None:
        try:
            span = trace.get_current_span()
            if span.is_recording():
                span.set_attribute(key, value)
        except Exception:
            pass

    # Scalar values

    @staticmethod
    def set(key: str, value: Optional[bool]) -> None:
        if key is None or value is None:
            return
        prefixed = MotadataDynamicInstrumentation._add_prefix(key)
        if prefixed:
            MotadataDynamicInstrumentation._try_set_attribute(prefixed, value)

    @staticmethod
    def set(key: str, value: Optional[int]) -> None:
        if key is None or value is None:
            return
        prefixed = MotadataDynamicInstrumentation._add_prefix(key)
        if prefixed:
            MotadataDynamicInstrumentation._try_set_attribute(prefixed, value)

    @staticmethod
    def set(key: str, value: Optional[float]) -> None:
        if key is None or value is None:
            return
        prefixed = MotadataDynamicInstrumentation._add_prefix(key)
        if prefixed:
            MotadataDynamicInstrumentation._try_set_attribute(prefixed, value)

    @staticmethod
    def set(key: str, value: Optional[str]) -> None:
        if key is None or value is None:
            return
        prefixed = MotadataDynamicInstrumentation._add_prefix(key)
        if prefixed:
            MotadataDynamicInstrumentation._try_set_attribute(prefixed, value)

    # List / Sequence values

    @staticmethod
    def set_bool_list(key: str, value: Optional[Sequence[bool]]) -> None:
        if not value or key is None or len(value) == 0:
            return
        prefixed = MotadataDynamicInstrumentation._add_prefix(key)
        if prefixed:
            try:
                MotadataDynamicInstrumentation._try_set_attribute(prefixed, list(value))
            except Exception:
                pass

    @staticmethod
    def set_int_list(key: str, value: Optional[Sequence[int]]) -> None:
        if not value or key is None or len(value) == 0:
            return
        prefixed = MotadataDynamicInstrumentation._add_prefix(key)
        if prefixed:
            try:
                MotadataDynamicInstrumentation._try_set_attribute(prefixed, list(value))
            except Exception:
                pass

    @staticmethod
    def set_float_list(key: str, value: Optional[Sequence[float]]) -> None:
        if not value or key is None or len(value) == 0:
            return
        prefixed = MotadataDynamicInstrumentation._add_prefix(key)
        if prefixed:
            try:
                MotadataDynamicInstrumentation._try_set_attribute(prefixed, list(value))
            except Exception:
                pass

    @staticmethod
    def set_string_list(key: str, value: Optional[Sequence[str]]) -> None:
        if not value or key is None or len(value) == 0:
            return
        prefixed = MotadataDynamicInstrumentation._add_prefix(key)
        if prefixed:
            try:
                MotadataDynamicInstrumentation._try_set_attribute(prefixed, list(value))
            except Exception:
                pass